src/OpenTelemetry/Metrics/Exemplar/FixedSizeExemplarReservoir.cs

```diff
@@ -0,0 +1,89 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Metrics;
+
+internal abstract class FixedSizeExemplarReservoir : ExemplarReservoir
+{
+    private readonly Exemplar[] runningExemplars;
+    private readonly Exemplar[] snapshotExemplars;
+
+    protected FixedSizeExemplarReservoir(int capacity)
+    {
+        Guard.ThrowIfOutOfRange(capacity, min: 1);
+
+        this.runningExemplars = new Exemplar[capacity];
+        this.snapshotExemplars = new Exemplar[capacity];
+        this.Capacity = capacity;
+    }
+
+    internal int Capacity { get; }
+
+    /// <summary>
+    /// Collects all the exemplars accumulated by the Reservoir.
+    /// </summary>
+    /// <returns><see cref="ReadOnlyExemplarCollection"/>.</returns>
+    public sealed override ReadOnlyExemplarCollection Collect()
+    {
+        if (this.ResetOnCollect)
+        {
+            for (int i = 0; i < this.runningExemplars.Length; i++)
```

Comment:
This could be simplified by moving the if condition inside:

```csharp
for (int i = 0; i < this.runningExemplars.Length; i++)
{
    ref var running = ref this.runningExemplars[i];
    if (running.IsUpdated())
    {
        running.Copy(ref this.snapshotExemplars[i]);

        if (this.ResetOnCollect)
        {
            running.Reset();
        }
    }
    else
    {
        this.snapshotExemplars[i].Reset();
    }
}
```

Reply:
Simple yes but adds more logic in the inner loop which slows it down.